### PR TITLE
Get rid of cache symlinks.

### PR DESCRIPTION
**Pull request by reitermarkus**
### Changes to the core
- [x] Followed [hacking.md](https://github.com/caskroom/homebrew-cask/blob/master/doc/development/hacking.md).

---
### Changes

The new cache directory is now at `HOMEBREW_CACHE`/`Cask`.

All symlinks in the old cache, `HOMEBREW_CACHE`/`Casks`, are checked, their sources moved to the new location, and then deleted. After this, `HOMEBREW_CACHE`/`Casks` is removed recursively.

When moving old cache files, they are also renamed according to a new naming convention, `#{token}--#{version}.#{ext}`, to allow for easier extraction of the token. The regular expression used for this should work for most cases, if it doesn't the worst case is that the file has to be re-downloaded.

---
### Todo
- [x] ~~Update Documentation~~ (there is none)

---

Fixes https://github.com/caskroom/homebrew-cask/issues/22727.

### CODE REVIEW

**Comment by reitermarkus** [comment]
@caskroom/maintainers, could you take a look?

**Comment by vitorgalvao** [comment]
What documentation do you envision needs updating? I don’t think we mention this anywhere.

**Comment by reitermarkus** [comment]
Right, I didn't actually check, just wanted to make sure I didn't forget to check.

**Comment by reitermarkus** [comment]
@caskroom/maintainers, anything that needs changing?

*Note:* `cache_completes` and `cache_incompletes` are not actually used anywhere, but I'll keep them in because of https://github.com/caskroom/homebrew-cask/issues/21763#issuecomment-224320234.

File: lib/hbc.rb [hunk 1]
```diff
@@ -51,8 +49,33 @@ module Hbc
...
+    if Hbc.legacy_cache.exist?
+      ohai "Migrating #{HOMEBREW_CACHE} to #{Hbc.cache}..."
+
+      relative_to_old_cache = Hbc.cache.relative_path_from(HOMEBREW_CACHE)
+
+      Hbc.legacy_cache.children.select(&:symlink?).each do |symlink|
+        file = symlink.readlink
+
+        new_name = file.basename
+                       .sub(%r{\-((?:(?:\d|#{Hbc::DSL::Version::DIVIDER_REGEX})+\-(?:\d|#{Hbc::DSL::Version::DIVIDER_REGEX})+)*[^\-]+)$},
```
Comment (jawshooah) [inline lib/hbc.rb]: Any way to simplify this substitution, or split it into several different ones? You could even split it into multiple lines with the `x` option to make it more readable.

**Comment by jawshooah** [comment]
Could you summarize the changes to `HOMEBREW_CACHE`, `Hbc.cache`, etc. in the top post? This PR does a bit more than get rid of cache symlinks.

File: lib/hbc.rb (hunk 1 above)
Comment (reitermarkus) [inline lib/hbc.rb]: Will split it with the `x` option. Unless you can come up with a simpler regex to rename `#{token}-#{version}.#{ext}` to `#{token}--#{version}.#{ext}`, that is.

Comment (jawshooah) [inline lib/hbc.rb]: Why are we replacing one hyphen with two? If everything is cached in a `Cask` subdirectory there shouldn't be any conflicts with cached kegs.

**Comment by reitermarkus** [comment]
@jawshooah, added some info on the changes made to the top post.

File: lib/hbc.rb (hunk 1 above)
Comment (jawshooah) [inline lib/hbc.rb]: Ah, just saw your addition to the top post, that makes sense.

Comment (reitermarkus) [inline lib/hbc.rb]: Also, I'd like to implement [what you suggested in this comment](https://github.com/caskroom/homebrew-cask/issues/21763#issuecomment-224320234), which is much easier if we know exactly where the token ends.

File: lib/hbc.rb [hunk 2]
```diff
@@ -51,8 +49,33 @@ module Hbc
 
   def self.init
     odebug "Creating directories"
-    HOMEBREW_CACHE.mkpath unless HOMEBREW_CACHE.exist?
-    HOMEBREW_CACHE_CASKS.mkpath unless HOMEBREW_CACHE_CASKS.exist?
+
+    # cache
+    Hbc.cache.mkpath unless Hbc.cache.exist?
+    if Hbc.legacy_cache.exist?
+      ohai "Migrating #{HOMEBREW_CACHE} to #{Hbc.cache}..."
```
Comment (jawshooah) [inline lib/hbc.rb]: I'd change the wording here to `Migrating cached Casks in #{HOMEBREW_CACHE} to #{Hbc.cache}...` since we're not migrating all of `HOMEBREW_CACHE`.

Comment (reitermarkus) [inline lib/hbc.rb]: Maybe simply `Migrating cached files to #{Hbc.cache}...`?

Comment (jawshooah) [inline lib/hbc.rb]: SGTM 👍

**Comment by mwean** [comment]
Looks good to me, nice cleanup! 👍

**Comment by reitermarkus** [comment]
This pull request was merged in 787593a2291460b837d59a3c1cf7129a9c5618f2...92f160f7160c365d3471ddbf68d2225dd6b7756c.
